internal/pkg/graph/graph_test.go

```diff
@@ -91,3 +92,157 @@ func TestGraph_IsAcyclic(t *testing.T) {
 		})
 	}
 }
+
+func TestGraph_Roots(t *testing.T) {
+	testCases := map[string]struct {
+		graph *Graph[int]
+
+		wantedRoots []int
+	}{
+		"should return nil if the graph is empty": {
+			graph: New[int](),
+		},
+		"should return all the vertices if there are no edges in the graph": {
+			graph:       New[int](1, 2, 3, 4, 5),
+			wantedRoots: []int{1, 2, 3, 4, 5},
+		},
+		"should return only vertices with no in degrees": {
+			graph: func() *Graph[int] {
+				g := New[int]()
+				g.Add(Edge[int]{
+					From: 1,
+					To:   3,
+				})
+				g.Add(Edge[int]{
+					From: 2,
+					To:   3,
+				})
+				g.Add(Edge[int]{
+					From: 3,
+					To:   4,
+				})
+				return g
+			}(),
+
+			wantedRoots: []int{1, 2},
+		},
+	}
+
+	for name, tc := range testCases {
+		t.Run(name, func(t *testing.T) {
+			require.ElementsMatch(t, tc.wantedRoots, tc.graph.Roots())
+		})
+	}
+}
+
+func TestLevelOrderTraversal(t *testing.T) {
+	testCases := map[string]struct {
+		graph *Graph[string]
+
+		wantedRanks     map[string]int
+		wantedErrPrefix string
+	}{
+		"should return an error when a cycle is detected": {
+			graph: &Graph[string]{
+				vertices: map[string]neighbors[string]{
+					"frontend": map[string]bool{
+						"backend": true,
+					},
+					"backend": map[string]bool{
+						"frontend": true,
+					},
+				},
+			},
```

Comment:
what do you think of creating the graph using the public API instead? like in `TestGraph_Roots()`